### DISCUSSION TITLE
Reset schema when `validateDatabaseSchema()` fails

### DISCUSSION THREAD

**Discussion by hiroshihorie** [simolus3/drift#2684] [Q&A] [answered]
I use Drift as a local cache for my app and I can afford to clear / recreate all data when schema changes.
How can I reset schema at launch if `validateDatabaseSchema()` fails and continue to run the app ?

**Comment by simolus3** [comment, marked as answer]
Something like this in the `MigrationStrategy` should work:

```dart
}, beforeOpen: (details) async {
  try {
    await validateDatabaseSchema();
  } on SchemaMismatch {
    // Delete all databases and re-create
    final migrator = createMigrator();
    final reversedEntities = allSchemaEntities.toList().reversed;

    for (final entity in reversedEntities) {
      await migrator.drop(entity);
    }

    await migrator.createAll();
  }
});
```

You don't need an `onUpgrade` callback at all in that case.

**Reply by hiroshihorie** [reply]
Thanks! But unfortunately I found out that `validateDatabaseSchema()` might not work with Web, am I correct ?
If so, do you have an alternative idea ?

**Reply by simolus3** [reply]
Yeah, unfortunately `validateDatabaseSchema()` does not currently work with web databases because it uses the native sqlite3 APIs to create another database to compare the two.

>  If so, do you have an alternative idea ?

You could raise the `schemaVersion` getter after making changes and then use `MigrationStrategy get migration => destructiveFallback`, which resets the database when `onUpgrade` is called.

**Reply by hiroshihorie** [reply]
Thanks, so I guess I will still need to bump the schemaVersion manually.

**Reply by albertmoravec** [reply]
Just to clarify, this would only drop entities currently present in the schema, right? If I removed a table from the schema and called this code afterwards without proper migration, it would still be present?

**Reply by simolus3** [reply]
Oh right, good catch! Yes, it would only delete tables that drift is aware of. A better approach may be to to close+delete the whole database before re-opening it, but that's not really possible from the `beforeOpen` callback. Another option may be to run a select on `select * from sqlite_schema` and use that to delete everything.

**Reply by albertmoravec** [reply]
I'm personally not in the exact same situation, so I decided to delete tables manually based on `sqlite_schema`. Deleting the whole database file would be the best option for me, but it is not currently possible for `WasmDatabase` AFAIK(?)

**Reply by simolus3** [reply]
You can use `WasmDatabase.probe` followed by `WasmProbeResult.deleteDatabase`, but it's not something that could be done within a migration callback.
